http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java

```diff
@@ -337,13 +354,27 @@ private ApacheHttpRequestConfig createRequestConfig(DefaultBuilder builder,
          * closed. This will not close connections currently in use. By default, this is enabled.
          */
         Builder useIdleConnectionReaper(Boolean useConnectionReaper);
+
+        /**
+         * Configuration that defines an HTTP route planner that computes the route an HTTP request should take.
+         * Overrides {@link #proxyConfiguration(ProxyConfiguration)}.
+         */
+        Builder httpRoutePlanner(HttpRoutePlanner proxyConfiguration);
+
+        /**
+         * Configuration that defines a custom credential provider for HTTP requests.
+         * Overrides {@link ProxyConfiguration#username()} and {@link ProxyConfiguration#password()}.
+         */
+        Builder credentialsProvider(CredentialsProvider credentialsProvider);
     }
 
     private static final class DefaultBuilder implements Builder {
         private final AttributeMap.Builder standardOptions = AttributeMap.builder();
         private ProxyConfiguration proxyConfiguration = ProxyConfiguration.builder().build();
         private InetAddress localAddress;
         private Boolean expectContinueEnabled;
+        private HttpRoutePlanner httpRoutePlanner;
+        private CredentialsProvider credentialsProvider;
```

Comment:
Thoughts on if we could do a generic version of this? Seems relatively simple.

Reply:
And by this I mean the CredentialsProvider

Reply:
There doesnt seem to be a good way...it becomes a rabbit hole of interfaces